RecoMET/METFilters/plugins/BadChargedCandidateFilter.cc

```diff
@@ -0,0 +1,159 @@
+  , taggingMode_          ( iConfig.getParameter<bool>    ("taggingMode") )
+  , debug_                ( iConfig.getParameter<bool>    ("debug") )
+  , maxDR_                ( iConfig.getParameter<double>  ("maxDR") )
+  , minPtDiffRel_         ( iConfig.getParameter<double>  ("minPtDiffRel") )
+  , minMuonTrackRelErr_   ( iConfig.getParameter<double>  ("minMuonTrackRelErr") )
+  , innerTrackRelErr_     ( iConfig.getParameter<double>  ("innerTrackRelErr") )
+  , minMuonPt_            ( iConfig.getParameter<double>  ("minMuonPt") )
+  , segmentCompatibility_ ( iConfig.getParameter<double>  ("segmentCompatibility") )
+{
+  produces<bool>();
+}
+
+BadChargedCandidateFilter::~BadChargedCandidateFilter() { }
+
+
+//
+// member functions
+//
+
+// ------------ method called on each new Event  ------------
+bool
+BadChargedCandidateFilter::filter(edm::StreamID iID, edm::Event& iEvent, const edm::EventSetup& iSetup) const
+{
+  using namespace std;
+  using namespace edm;
+
+  typedef View<reco::Candidate> CandidateView;
+  Handle<CandidateView> pfCandidates;
+  iEvent.getByToken(tokenPFCandidates_,pfCandidates);
+
+  typedef View<reco::Muon> MuonView;
+  Handle<MuonView> muons;
+  iEvent.getByToken(tokenMuons_,muons);
+
+  bool foundBadChargedCandidate = false;
+
+  for ( unsigned i=0; i < muons->size(); ++i ) { // loop over all muons
+
+    const reco::Muon & muon = (*muons)[i];
+
+    reco::TrackRef bestMuonTrack = muon.muonBestTrack();
+
+    if (debug_) cout<<"BadChargedCandidate test:Muon "<< i << endl;
+    // reco::TrackRef innerMuonTrack = muon.innerTrack();
+    //  if ( muon.pt() < minMuonPt_ && innerMuonTrack->pt() < minMuonPt_) {
+    //  if (debug_) cout <<"skipping the muon because low muon pt" << endl;
+    //  continue ; } {
+    {
+       reco::TrackRef innerMuonTrack = muon.innerTrack();
+        if (debug_) cout<<"muon "<<muon.pt()<<endl;
+	
+        if ( innerMuonTrack.isNull() ) { 
+            if (debug_) cout<<"Skipping this muon because it has no inner track"<<endl; 
+            continue; 
+            };
+
+	if ( muon.pt() < minMuonPt_ && innerMuonTrack->pt() < minMuonPt_) {
+          if (debug_) cout <<"skipping the muon because low muon pt" << endl;
+          continue;
+        }
+
+	// Consider only Global Muons  	
+	if (muon.isGlobalMuon() == 0) {
+	  if(debug_) cout << "Skipping this muon because not a Global Muon" << endl;
+	  continue;
+	}
+
+	if (debug_) cout << "SegmentCompatibility :"<< muon::segmentCompatibility(muon) << "RelPtErr:" << bestMuonTrack->ptError()/bestMuonTrack->pt() << endl;
+	if (muon::segmentCompatibility(muon) > segmentCompatibility_ && bestMuonTrack->ptError()/bestMuonTrack->pt() < minMuonTrackRelErr_ && innerMuonTrack->ptError()/innerMuonTrack->pt() < innerTrackRelErr_ ) {
+	  if (debug_) cout <<"Skipping this muon because segment compatiblity > 0.3 and relErr(best track) <2 and relErr(inner track) < 1 " << endl;
```

Comment:
the printouts should match the cut values (``` 0.3 ``` vs actual value of segmentCompatibility_)